From 3557ab28476b38f0734560a6464102c58efe603c Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Tue, 29 Oct 2024 14:42:39 +0800
Subject: [PATCH 255/277] fixup! mtk: mt76: mt7996: Refactor handling vht
 operating mode notification

Fix incorrect operating-mode setting during STA addition.
Specifically, STA_REC_RA command provides operating-mode fields, which were not used.

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 mt7996/mcu.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index d0b89c1c..e4f8ff97 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3146,9 +3146,11 @@ mt7996_mcu_sta_rate_ctrl_tlv(struct sk_buff *skb, struct mt7996_dev *dev,
 			     struct ieee80211_link_sta *link_sta)
 {
 #define INIT_RCPI 180
+	enum ieee80211_sta_rx_bandwidth cap_bw = ieee80211_link_sta_cap_bw(link_sta);
 	struct mt76_phy *mphy = mconf->phy->mt76;
 	struct cfg80211_chan_def *chandef = &mphy->chandef;
 	struct cfg80211_bitrate_mask *mask = &mconf->bitrate_mask;
+	u8 cap_nss = ieee80211_link_sta_cap_nss(link_sta);
 	enum nl80211_band band = chandef->chan->band;
 	struct sta_rec_ra_uni *ra;
 	struct tlv *tlv;
@@ -3162,10 +3164,13 @@ mt7996_mcu_sta_rate_ctrl_tlv(struct sk_buff *skb, struct mt7996_dev *dev,
 	ra->auto_rate = true;
 	ra->phy_mode = mt76_connac_get_phy_mode(mphy, conf->vif, band, link_sta);
 	ra->channel = chandef->chan->hw_value;
-	ra->bw = (link_sta->bandwidth == IEEE80211_STA_RX_BW_320) ?
-		 CMD_CBW_320MHZ : link_sta->bandwidth;
-	ra->phy.bw = ra->bw;
+	ra->bw = mt76_connac_chan_bw(chandef);
+	ra->phy.bw = (cap_bw == IEEE80211_STA_RX_BW_320) ? CMD_CBW_320MHZ : cap_bw;
 	ra->mmps_mode = mt7996_mcu_get_mmps_mode(link_sta->smps_mode);
+	ra->op_mode = cap_bw != link_sta->bandwidth || cap_nss != link_sta->rx_nss;
+	ra->op_vht_chan_width = link_sta->bandwidth;
+	ra->op_vht_rx_nss = link_sta->rx_nss - 1;
+	ra->op_vht_rx_nss_type = 0;
 
 	if (supp_rate) {
 		supp_rate &= mask->control[band].legacy;
-- 
2.45.2

